Failed with:

```
10:39:41  [ERROR]   WebappClassLoaderTest.getURLs:116 » InaccessibleObject Unable to make field private final java.lang.ClassLoader sun.rmi.transport.Target.ccl accessible: module java.rmi does not "opens sun.rmi.transport" to unnamed module @314c508a
10:39:41  [INFO] 
10:39:41  [ERROR] Tests run: 4, Failures: 0, Errors: 1, Skipped: 0
```

--- a/appserver/web/war-util/src/main/java/org/glassfish/web/loader/ReferenceCleaner.java
+++ b/appserver/web/war-util/src/main/java/org/glassfish/web/loader/ReferenceCleaner.java
@@ -23,7 +23,6 @@ import java.lang.System.Logger;
 import java.lang.ref.Reference;
 import java.lang.reflect.AccessibleObject;
 import java.lang.reflect.Field;
-import java.lang.reflect.InaccessibleObjectException;
 import java.lang.reflect.Method;
 import java.lang.reflect.Modifier;
 import java.security.AccessController;
@@ -32,14 +31,12 @@ import java.sql.Driver;
 import java.sql.DriverManager;
 import java.util.Collection;
 import java.util.Iterator;
-import java.util.Map;
 import java.util.ResourceBundle;
 
 import org.glassfish.web.util.IntrospectionUtils;
 
 import static java.lang.System.Logger.Level.DEBUG;
 import static java.lang.System.Logger.Level.ERROR;
-import static java.lang.System.Logger.Level.INFO;
 import static java.lang.System.Logger.Level.TRACE;
 import static java.lang.System.Logger.Level.WARNING;
 import static java.text.MessageFormat.format;
@@ -75,7 +72,6 @@ class ReferenceCleaner {
 
             clearReferencesJdbc();
             checkThreadLocalsForLeaks();
-            clearReferencesRmiTargets();
 
             // Null out any static or final fields from loaded classes,
             // as a workaround for apparent garbage collection bugs
@@ -213,67 +209,6 @@ class ReferenceCleaner {
             if (valueLeak) {
                 LOG.log(ERROR, CHECK_THREAD_LOCALS_FOR_LEAKS, loader.getName(), keyDescription, describe(value));
             }
-        }
-    }
-
-
-    /**
-     * Clear RMI Targets loaded by this class loader.
-     * This depends on the internals of the JVM so it does everything by reflection.
-     */
-    private void clearReferencesRmiTargets() {
-        LOG.log(TRACE, "clearReferencesRmiTargets()");
-        try {
-            // Need access to the ccl field of sun.rmi.transport.Target
-            Class<?> objectTargetClass = Class.forName("sun.rmi.transport.Target");
-            Field cclField = objectTargetClass.getDeclaredField("ccl");
-            setAccessible(cclField);
-
-            // Clear the objTable map
-            Class<?> objectTableClass = Class.forName("sun.rmi.transport.ObjectTable");
-            Field objTableField = objectTableClass.getDeclaredField("objTable");
-            setAccessible(objTableField);
-            Object objTable = objTableField.get(null);
-            if (objTable == null) {
-                return;
-            }
-
-            // Iterate over the values in the table
-            if (objTable instanceof Map<?, ?>) {
-                Iterator<?> iter = ((Map<?, ?>) objTable).values().iterator();
-                while (iter.hasNext()) {
-                    Object obj = iter.next();
-                    Object cclObject = cclField.get(obj);
-                    if (loader == cclObject) {
-                        iter.remove();
-                    }
-                }
-            }
-
-            // Clear the implTable map
-            Field implTableField = objectTableClass.getDeclaredField("implTable");
-            setAccessible(implTableField);
-            Object implTable = implTableField.get(null);
-            if (implTable == null) {
-                return;
-            }
-
-            // Iterate over the values in the table
-            if (implTable instanceof Map<?, ?>) {
-                Iterator<?> iter = ((Map<?, ?>) implTable).values().iterator();
-                while (iter.hasNext()) {
-                    Object obj = iter.next();
-                    Object cclObject = cclField.get(obj);
-                    if (loader == cclObject) {
-                        iter.remove();
-                    }
-                }
-            }
-        } catch (ClassNotFoundException e) {
-            LOG.log(INFO, getString(LogFacade.CLEAR_RMI_INFO, loader.getName()), e);
-        } catch (SecurityException | NoSuchFieldException | IllegalArgumentException | IllegalAccessException
-            | InaccessibleObjectException e) {
-            LOG.log(WARNING, getString(LogFacade.CLEAR_RMI_FAIL, loader.getName()), e);
         }
     }
 

--- a/appserver/web/war-util/src/main/java/org/glassfish/web/loader/ReferenceCleaner.java
+++ b/appserver/web/war-util/src/main/java/org/glassfish/web/loader/ReferenceCleaner.java
@@ -23,6 +23,7 @@ import java.lang.System.Logger;
 import java.lang.ref.Reference;
 import java.lang.reflect.AccessibleObject;
 import java.lang.reflect.Field;
+import java.lang.reflect.InaccessibleObjectException;
 import java.lang.reflect.Method;
 import java.lang.reflect.Modifier;
 import java.security.AccessController;
@@ -270,7 +271,8 @@ class ReferenceCleaner {
             }
         } catch (ClassNotFoundException e) {
             LOG.log(INFO, getString(LogFacade.CLEAR_RMI_INFO, loader.getName()), e);
-        } catch (SecurityException | NoSuchFieldException | IllegalArgumentException | IllegalAccessException e) {
+        } catch (SecurityException | NoSuchFieldException | IllegalArgumentException | IllegalAccessException
+            | InaccessibleObjectException e) {
             LOG.log(WARNING, getString(LogFacade.CLEAR_RMI_FAIL, loader.getName()), e);
         }
     }

--- a/appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoader.java
+++ b/appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoader.java
@@ -998,7 +998,9 @@ public final class WebappClassLoader extends GlassfishUrlClassLoader
 
     @Override
     public synchronized URL[] getURLs() {
-        checkStatus(LifeCycleStatus.RUNNING);
+        if (this.status == LifeCycleStatus.CLOSED) {
+            return new URL[0];
+        }
         if (repositoryURLs != null) {
             return repositoryURLs.toArray(URL[]::new);
         }

--- a/appserver/web/war-util/src/test/java/org/glassfish/web/loader/WebappClassLoaderTest.java
+++ b/appserver/web/war-util/src/test/java/org/glassfish/web/loader/WebappClassLoaderTest.java
@@ -18,6 +18,7 @@ package org.glassfish.web.loader;
 
 import com.sun.enterprise.util.io.FileUtils;
 
+import java.io.IOException;
 import java.net.URL;
 import java.net.URLClassLoader;
 import java.nio.file.Files;
@@ -25,11 +26,16 @@ import java.nio.file.Path;
 
 import org.apache.naming.resources.WebDirContext;
 import org.glassfish.common.util.GlassfishUrlClassLoader;
+import org.hamcrest.core.StringEndsWith;
 import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
 
 import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.arrayContaining;
+import static org.hamcrest.Matchers.arrayWithSize;
+import static org.hamcrest.Matchers.hasLength;
+import static org.hamcrest.Matchers.stringContainsInOrder;
 import static org.hamcrest.core.StringEndsWith.endsWith;
 import static org.junit.jupiter.api.Assertions.assertAll;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
@@ -73,12 +79,7 @@ class WebappClassLoaderTest {
 
     @Test
     public void getResource() {
-        WebappClassLoader classLoader = new WebappClassLoader(WebappClassLoaderTest.class.getClassLoader());
-        WebDirContext webDirContext = new WebDirContext();
-        webDirContext.setDocBase(docBase.toFile().getAbsolutePath());
-        classLoader.setResources(webDirContext);
-        classLoader.addRepository(repo.getFileName().toString() + "/", repo.toFile());
-        classLoader.setDelegate(false);
+        WebappClassLoader classLoader = createCL();
         classLoader.start();
         assertAll(
             () -> assertThat(classLoader.getResource(".").toExternalForm(), endsWith("/repo1/")),
@@ -92,12 +93,7 @@ class WebappClassLoaderTest {
 
     @Test
     public void getResourceAsStream() {
-        WebappClassLoader classLoader = new WebappClassLoader(WebappClassLoaderTest.class.getClassLoader());
-        WebDirContext webDirContext = new WebDirContext();
-        webDirContext.setDocBase(docBase.toFile().getAbsolutePath());
-        classLoader.setResources(webDirContext);
-        classLoader.addRepository(repo.getFileName().toString() + "/", repo.toFile());
-        classLoader.setDelegate(false);
+        WebappClassLoader classLoader = createCL();
         classLoader.start();
         assertAll(
             // streams of 4 bytes returned by the delegate CL
@@ -108,5 +104,27 @@ class WebappClassLoaderTest {
             () -> assertNull(classLoader.getResourceAsStream("Dir1"), "Dir1"),
             () -> assertNotNull(classLoader.getResourceAsStream("Dir1/fileInDir1.txt"), "Dir1/fileInDir1.txt")
         );
+    }
+
+
+    @Test
+    public void getURLs() throws Exception {
+        WebappClassLoader classLoader = createCL();
+        assertThat(classLoader.getURLs(), arrayWithSize(1));
+        classLoader.start();
+        assertThat(classLoader.getURLs(), arrayWithSize(1));
+        classLoader.close();
+        assertThat(classLoader.getURLs(), arrayWithSize(0));
+    }
+
+
+    private WebappClassLoader createCL() {
+        WebappClassLoader classLoader = new WebappClassLoader(WebappClassLoaderTest.class.getClassLoader());
+        WebDirContext webDirContext = new WebDirContext();
+        webDirContext.setDocBase(docBase.toFile().getAbsolutePath());
+        classLoader.setResources(webDirContext);
+        classLoader.addRepository(repo.getFileName().toString() + "/", repo.toFile());
+        classLoader.setDelegate(false);
+        return classLoader;
     }
 }

